### PR TITLE
jwt_authn: Add header_to_metadata

### LINKED ISSUE #18138: No way to retrieve JWT kid as metadata
*Title*: No way to retrieve JWT tokens from metadata

*Description*:

The JWT option [payload_in_metadata](https://www.envoyproxy.io/docs/envoy/v1.19.1/api-v3/extensions/filters/http/jwt_authn/v3/config.proto#envoy-v3-api-field-extensions-filters-http-jwt-authn-v3-jwtprovider-payload-in-metadata) allows you to store the metadata stored in the JWT payload, for logging or further processing. However, JWT headers like [kid](https://www.rfc-editor.org/rfc/rfc7515.html#section-4.1.4) are not passed on, presumably because they are headers and not payloads.

*Why I care*:

Validating the payload in my use-case is useful because it forms part of our authentication chain: Every `kid` is valid for at most one `sub`, and so the `kid` needs to be checked to ensure a client isn't misrepresenting their `sub`.

*Note*:
"header" can be a bit confusing here. JWTs are often passed as `HTTP` headers, but JWTs themselves have headers, bits of data that aren't signed so they can be used for validating signatures. [jwt.io](https://jwt.io/#debugger-io) shows the difference between `header` and `payload` clearly.

[optional *Relevant Links*:] attached inline

### PR DESCRIPTION
**Pull request by dio**
Commit Message: 

This patch adds the `header_to_metadata` field to the `JwtProvider` config to allow setting the extracted header of a successfully verified JWT to dynamic metadata.

Additional Description:
Risk Level: Low, a new feature
Testing: Added
Docs Changes: Added
Release Notes: Added
Platform-Specific Features: N/A
Fixes #18138

### CODE REVIEW

**Comment by lizfeed** [comment]
Thanks for the quick response! I've verified I can log this new metadata field in this release. I use [envoy_data_plane](https://github.com/cetanu/envoy_data_plane), so I can also verify that the protobufs are generated correctly at least for python.

File: source/extensions/filters/http/jwt_authn/verifier.cc [hunk 1]
```diff
@@ -72,7 +82,8 @@ class ContextImpl : public Verifier::Context {
   Verifier::Callbacks* callback_;
   absl::node_hash_map<const Verifier*, CompletionState> completion_states_;
   std::vector<AuthenticatorPtr> auths_;
-  ProtobufWkt::Struct payload_;
+  ProtobufWkt::Struct jwt_payload_;
+  ProtobufWkt::Struct jwt_header_;
```
Comment (dio) [inline source/extensions/filters/http/jwt_authn/verifier.cc:86]: Will change this by using only one struct as per @yangminzhu suggestion.

File: api/envoy/extensions/filters/http/jwt_authn/v3/config.proto [hunk 2]
```diff
@@ -230,6 +230,40 @@ message JwtProvider {
   //
   string payload_in_metadata = 9;
 
+  // If non empty, similar to :ref:`payload <envoy_v3_api_field_extensions.filters.http.jwt_authn.v3.JwtProvider.payload_in_metadata>`,
```
Comment (yangminzhu) [inline api/envoy/extensions/filters/http/jwt_authn/v3/config.proto:233]: `payload` -> `payload_in_metadata`

File: api/envoy/extensions/filters/http/jwt_authn/v3/config.proto [hunk 3]
```diff
@@ -230,6 +230,40 @@ message JwtProvider {
...
+  //       iss: https://example.com
+  //       sub: test@example.com
+  //       aud: https://example.com
+  //       exp: 1501281058
+  //     my_header:
+  //       alg: JWT
+  //       kid: EF71iSaosbC5C4tC6Syq1Gm647M
+  //       alg: PS256
+  //
+  string header_in_metadata = 14;
```
Comment (yangminzhu) [inline api/envoy/extensions/filters/http/jwt_authn/v3/config.proto:265]: what if `payload_in_metadata` is the same as `header_in_metadata`? Is it allowed?

Comment (qiwzhang) [reply api/envoy/extensions/filters/http/jwt_authn/v3/config.proto:265]: it should not be allowed, please add comment on this restriction.

Comment (dio) [reply api/envoy/extensions/filters/http/jwt_authn/v3/config.proto:271]: Added a warning section. Please let me know if we want to validate this when checking the config.

File: api/envoy/extensions/filters/http/jwt_authn/v3/config.proto (hunk 2 above)
Comment (dio) [reply api/envoy/extensions/filters/http/jwt_authn/v3/config.proto]: Resolved. Thank you!

File: api/envoy/extensions/filters/http/jwt_authn/v3/config.proto (hunk 3 above)
Comment (yangminzhu) [reply api/envoy/extensions/filters/http/jwt_authn/v3/config.proto:271]: hmm, the current comment says `... is not allowed ...` which I would expect the validation is going to reject the config, or you can just update the comment to be more accurate (it's not suggested due to potential override but can still be used (won't be rejected in validation) if the user is sure it won't have any override for their use case).

File: source/extensions/filters/http/jwt_authn/authenticator.cc [hunk 4]
```diff
@@ -90,7 +90,7 @@ class AuthenticatorImpl : public Logger::Loggable<Logger::Id::jwt>,
   Http::HeaderMap* headers_{};
   // The active span for the request
   Tracing::Span* parent_span_{&Tracing::NullSpan::instance()};
-  // the callback function to set payload
+  // The callback function called to set the extracted payload and header from a verified JWT.
```
Comment (yangminzhu) [inline source/extensions/filters/http/jwt_authn/authenticator.cc:93]: `set_payload_cb_` should have a more generic name as it now sets both payload and header.

**Review by qiwzhang — APPROVED** [review]

**Comment by dio** [comment]
@qiwzhang @yangminzhu I have a question, re: https://github.com/envoyproxy/envoy/blob/a3cc673ccc6f2c42d92d63de11b1aa781c5f991d/source/extensions/filters/http/jwt_authn/verifier.h#L29-L34. Do you think this is still correct. I'm not sure I understand there is a use-case that we do this (storing the payload as a string).

File: api/envoy/extensions/filters/http/jwt_authn/v3/config.proto (hunk 3 above)
Comment (dio) [reply api/envoy/extensions/filters/http/jwt_authn/v3/config.proto:271]: @yangminzhu I updated as you suggested.

File: source/extensions/filters/http/jwt_authn/authenticator.cc (hunk 4 above)
Comment (dio) [reply source/extensions/filters/http/jwt_authn/authenticator.cc:94]: Updated.

**Comment by mattklein123** [comment]
Will wait for review/approve from @yangminzhu or @qiwzhang before looking, thanks.

/wait-any

**Review by qiwzhang — APPROVED** [review]

**Review by yangminzhu — APPROVED** [review]

**Comment by yangminzhu** [comment]
> @qiwzhang @yangminzhu I have a question, re:
> 
> https://github.com/envoyproxy/envoy/blob/a3cc673ccc6f2c42d92d63de11b1aa781c5f991d/source/extensions/filters/http/jwt_authn/verifier.h#L29-L34
> 
> . Do you think this is still correct? I'm not sure I understand there is a use-case when we do this (storing the payload as a string).

`storing the payload as a string` is to pass the payload to backend application without the signature so that the backend can not reuse the JWT token.

**Review by mattklein123 — APPROVED** [review]

**Pull request merged**